app/src/main/java/com/duckduckgo/app/httpsupgrade/api/HttpsUpgradeListDownloader.kt

```diff
@@ -18,41 +18,75 @@ package com.duckduckgo.app.httpsupgrade.api
 
 import com.duckduckgo.app.global.api.isCached
 import com.duckduckgo.app.global.db.AppDatabase
+import com.duckduckgo.app.httpsupgrade.db.HttpsUpgradeDbWriteStatusStore
 import com.duckduckgo.app.httpsupgrade.db.HttpsUpgradeDomainDao
 import io.reactivex.Completable
 import timber.log.Timber
 import java.io.IOException
 import javax.inject.Inject
 
+/**
+ * For performance reasons, we can't insert all the HTTPS rows in a single transaction as it blocks other DB writes from happening for a while
+ * To avoid the performance problem of a single transaction, we can chunk the data and perform many smaller transactions.
+ * However, if we don't use a single transaction, there is the risk the process could die or exception be thrown while iterating through the HTTPS inserts, leaving the DB in a bad state.
+ * The network cache could therefore deliver the cached value and we could skip writing to the DB thinking we'd already successfully written it all.
+ *
+ * To counter this, we need an external mechanism to track whether the write completed successfully or not.
+ * Using this, we can detect if the write failed when we next download the data. We'll only skip the DB inserts if cache returns the list and the
+ * the eTag received in the response headers matches the eTag we stored on last successful write, indicating that it finished writing successfully.
+ * @see HttpsUpgradeDbWriteStatusStore for how this status flag is stored.
+ */
 class HttpsUpgradeListDownloader @Inject constructor(
-        private val service: HttpsUpgradeListService,
-        private val database: AppDatabase,
-        private val httpsUpgradeDao: HttpsUpgradeDomainDao) {
+    private val service: HttpsUpgradeListService,
+    private val database: AppDatabase,
+    private val httpsUpgradeDao: HttpsUpgradeDomainDao,
+    private val dbWriteStatusStore: HttpsUpgradeDbWriteStatusStore
+) {
 
-    fun downloadList(): Completable {
+    fun downloadList(chunkSize: Int = INSERTION_CHUNK_SIZE): Completable {
 
         Timber.d("Downloading HTTPS Upgrade data")
 
         return Completable.fromAction {
 
             val call = service.https()
             val response = call.execute()
+            val eTag = response.headers().get("etag")
```

Comment:
constant for this? (optional)

Headers are case-insensitive - does this class still return the etag if the header comes back as say "ETag"? (it's kind of moot as we use etag in our HTTP requests)

Reply:
Added a constant.

It uses a `equalsIgnoreCare` under the hood